### PR TITLE
#165020094  Add Search Functionality and Custom filtering 

### PR DESCRIPTION
**Pull request by coolbeatz71**
#### What does this PR do?
Add the Search Filter feature
#### Description of Task to be completed?
- create the reducer
- create actions and action types
- connect SearchBox component with redux
- implement method to handle search filter
- add animation to the search button
- revamp integration test

#### How should this be manually tested?

- npm test

#### Any background context you want to provide?

- user can search for articles by `title`, `author`, `tag` or  any `keyword`

#### What are the relevant pivotal tracker stories?

[#165020094](https://www.pivotaltracker.com/story/show/165020094)

#### Screenshots (if appropriate)

N/A

#### Questions:

N/A

### CODE REVIEW

File: src/Components/Common/SearchBox/SearchBox.jsx [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
+  Button,
+  Spinner,
+} from 'react-bootstrap';
 import { PropTypes } from 'prop-types';
 import { faSearch } from '@fortawesome/free-solid-svg-icons';
 import 'bootstrap/dist/css/bootstrap.css';
 import './SearchBox.scss';
+import { search } from '../../../actions/search';
 
 const filters = ['tag', 'author', 'title'];
```
Comment (gracelungu) [inline src/Components/Common/SearchBox/SearchBox.jsx:18]: @coolbeatz71  Why are these filters out of the class?

File: src/Components/Common/SearchBox/SearchBox.jsx [hunk 2]
```diff
@@ -28,6 +37,55 @@ class SearchBox extends Component {
...
+          break;
+        case 'title':
+          url = `/api/search?title=${value}`;
+          break;
+        case 'tag':
+          url = `/api/search?tag=${value}`;
+          break;
+        default:
+          url = `/api/search?keyword=${value}`;
+      }
```
Comment (gracelungu) [inline src/Components/Common/SearchBox/SearchBox.jsx]: @coolbeatz71 This code can be written simply like this : `/api/search?{filter || 'keyword'}=${value}`

File: src/Components/Common/SearchBox/SearchBox.jsx [hunk 3]
```diff
@@ -28,6 +37,55 @@ class SearchBox extends Component {
...
+   * @param {object} e
+   * @returns {void}
+   */
+  onSubmitSearch = e => {
+    e.preventDefault();
+    let url;
+    const { value } = this.props;
+    const { filter } = this.state;
+    const { onSearch } = this.props;
+    const { history } = this.props;
```
Comment (gracelungu) [inline src/Components/Common/SearchBox/SearchBox.jsx]: @coolbeatz71 Destructuring from props could have been done once.

File: src/Containers/Search/NotFound.jsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import Image from 'react-bootstrap/Image';
+import 'bootstrap/dist/css/bootstrap.css';
+import imgNotFound from '../../assets/images/undraw_not_found.svg';
+
+const imgStyle = {
+  width: '45%',
+};
```
Comment (gracelungu) [inline src/Containers/Search/NotFound.jsx:8]: @coolbeatz71 Avoid writing plain CSS, in the component.

File: src/Containers/Search/ResultCard.jsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    </Card.Header>
+    <Card.Body className="d-flex flex-row">
+      <div>
+        <Card.Img variant="top" className="article-img" src={articleImg} />
+      </div>
+      <div className="article-info">
+        <Card.Title className="title">{title}</Card.Title>
+        <Card.Subtitle className="mb-2 subtitle">{description}</Card.Subtitle>
+        <a href={`${paths.ARTICLES_PATH}/${slug}`}>
+          <Button size={12} text="View more..." />
```
Comment (gracelungu) [inline src/Containers/Search/ResultCard.jsx:35]: @coolbeatz71 There is no action attached to this button

File: src/Containers/Search/SearchResult.jsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  return (
+    <React.Fragment>
+      <NavBar {...props} />
+      {articles.length === 0 ? (
+        <NotFound />
+      ) : (
+        <div className="d-flex flex-column results">
+          <Col md={12} sm={12} lg={8} className="result-column">
+            {articles.map(item => {
```
Comment (gracelungu) [inline src/Containers/Search/SearchResult.jsx:31]: @coolbeatz71 Destructuring is preferred here, to different item fields.

File: src/Components/Common/SearchBox/SearchBox.jsx (hunk 1 above)
Comment (coolbeatz71) [reply src/Components/Common/SearchBox/SearchBox.jsx:18]: Yes @gracelungu, the filters array doesn't need to be part of the class. I am only looping through this array in order to display each `filter type` on the search box dropdown

File: src/Components/Common/SearchBox/SearchBox.jsx (hunk 2 above)
Comment (coolbeatz71) [reply src/Components/Common/SearchBox/SearchBox.jsx]: Wow, @gracelungu good idea brother. I am refactoring now. Thx

File: src/Components/Common/SearchBox/SearchBox.jsx (hunk 3 above)
Comment (coolbeatz71) [reply src/Components/Common/SearchBox/SearchBox.jsx]: @gracelungu thx for the feedback, I am fixing the issue ASAP

File: src/Containers/Search/ResultCard.jsx (hunk 5 above)
Comment (coolbeatz71) [reply src/Containers/Search/ResultCard.jsx:35]: @gracelungu, thank you for the feedback, I would like to tell you that ```<a href={`${paths.ARTICLES_PATH}/${slug}`}>``` is a link which should display the specific article by the slug. This is the action in question

File: src/Containers/Search/SearchResult.jsx (hunk 6 above)
Comment (coolbeatz71) [reply src/Containers/Search/SearchResult.jsx:31]: Yes @gracelungu you are right

**Review by gracelungu — APPROVED** [review]

**Review by giseleiradu — APPROVED** [review]

**Pull request merged**